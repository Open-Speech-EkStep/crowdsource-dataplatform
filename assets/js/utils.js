const { HOUR_IN_SECONDS, SIXTY } = require("./constants");

const ALL_LANGUAGES = [
    {value: "Assamese",id: "as", text: "অসমীয়া", hasLocaleText: true},
    {value: "Bengali", id: "bn", text: "বাংলা", hasLocaleText: true},
    {value: "English", id: "en", text: "English", hasLocaleText: true},
    {value: "Gujarati", id: "gu", text: "ગુજરાતી", hasLocaleText: true},
    {value: "Hindi", id: "hi", text: "हिंदी", hasLocaleText: true},
    {value: "Kannada", id: "kn", text: "ಕನ್ನಡ", hasLocaleText: true},
    {value: "Malayalam", id: "ml", text: "മലയാളം", hasLocaleText: true},
    {value: "Marathi", id: "mr", text: "मराठी", hasLocaleText: true},
    {value: "Odia", id: "or", text: "ଓଡିଆ", hasLocaleText: true},
    {value: "Punjabi", id: "pa", text: "ਪੰਜਾਬੀ", hasLocaleText: true},
    {value: "Tamil", id: "ta", text: "தமிழ்", hasLocaleText: true},
    {value: "Telugu", id: "te", text: "తెలుగు", hasLocaleText: true}];

function convertPXToVH(px) {
    return px * (100 / document.documentElement.clientHeight);
}

function setPageContentHeight() {
    const $footer = $('footer');
    const $nav = $('.navbar');
    const edgeHeightInPixel = $footer.outerHeight() + $nav.outerHeight()
    const contentHeightInVH = 100 - convertPXToVH(edgeHeightInPixel)
    $('#content-wrapper').css('min-height', contentHeightInVH + 'vh');
}

function toggleFooterPosition() {
    const $footer = $('footer');
    $footer.toggleClass('fixed-bottom')
    $footer.toggleClass('bottom')
}

function fetchLocationInfo() {
    //https://api.ipify.org/?format=json
    return fetch('https://www.cloudflare.com/cdn-cgi/trace').then(res => res.text()).then(ipAddressText => {
        const dataArray = ipAddressText.split('\n');
        let ipAddress = "";
        for (let ind in dataArray) {
            if (dataArray[ind].startsWith("ip=")) {
                ipAddress = dataArray[ind].replace('ip=', '');
                break;
            }
        }
        if (ipAddress.length !== 0) {
            return fetch(`/location-info?ip=${ipAddress}`);
        } else {
            return new Promise((resolve, reject) => {
                reject("Ip Address not available")
            })
        }
    });
}

const performAPIRequest = (url) => {
    return fetch(url).then((data) => {
        if(!data.ok) {
            throw Error(data.statusText || 'HTTP error');
        } else {
            return Promise.resolve(data.json());
        }
    });
}

const getLocaleString = function() {
    performAPIRequest('/get-locale-strings')
        .then((response) => {
            localStorage.setItem('localeString', JSON.stringify(response));
        });
}

const updateLocaleLanguagesDropdown = (language) => {
    const dropDown = $('#localisation_dropdown');
    const localeLang = ALL_LANGUAGES.find(ele => ele.value === language);
    if (language.toLowerCase() === "english" || localeLang.hasLocaleText === false) {
        dropDown.html('<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>');
    } else {
        dropDown.html(`<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>
        <a id=${localeLang.value} class="dropdown-item" href="/changeLocale/${localeLang.id}">${localeLang.text}</a>`);
    }
}

const calculateTime = function (totalSeconds, isSeconds = true) {
    const hours = Math.floor(totalSeconds / HOUR_IN_SECONDS);
    const remainingAfterHours = totalSeconds % HOUR_IN_SECONDS;
    const minutes = Math.floor(remainingAfterHours / SIXTY);
    const seconds = Math.round(remainingAfterHours % SIXTY);
    if (isSeconds) {
        return { hours, minutes, seconds };
    } else {
        return { hours, minutes };
    }
};

const formatTime = function (hours, minutes = 0, seconds = 0) {
    let result = "";
    if (hours > 0) {
        result += `${hours} hrs `;
    }
    if (minutes > 0) {
        result += `${minutes} min `;
    }
    if (hours === 0 && minutes === 0 && seconds > 0) {
        result += `${seconds} sec `;
    }
    return result.substr(0, result.length - 1);
};

module.exports = { setPageContentHeight, toggleFooterPosition, fetchLocationInfo, updateLocaleLanguagesDropdown ,calculateTime, formatTime, getLocaleString, performAPIRequest}
