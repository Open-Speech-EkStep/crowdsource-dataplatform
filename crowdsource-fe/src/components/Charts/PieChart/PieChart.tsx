import { useRef, useEffect } from 'react';

import * as am4charts from '@amcharts/amcharts4/charts';
import * as am4core from '@amcharts/amcharts4/core';

interface ChartProps {
  id: string;
  data: {
    data: Array<any>;
    colors?: Array<string>;
    isScrollbar?: boolean;
    tooltipTemplate?: string;
    xAxisLabel?: string | any;
    yAxisLabel?: string | any;
  };
}
const PieChart = (props: ChartProps | any) => {
  const chart = useRef({});

  useEffect(() => {
    var x = am4core.create(props.id, am4charts.PieChart3D);
    x.hiddenState.properties.opacity = 0;

    x.data = props.data.data;

    x.innerRadius = am4core.percent(40);
    x.depth = 20;

    x.legend = new am4charts.Legend();

    var series = x.series.push(new am4charts.PieSeries3D());
    series.dataFields.value = 'value';
    series.labels.template.disabled = true;
    series.ticks.template.disabled = true;
    series.dataFields.depthValue = 'value';
    series.dataFields.category = 'category';
    series.slices.template.cornerRadius = 5;
    // series.colors.step = 3;

    /* istanbul ignore next */
    series.slices.template.adapter.add('fill', (value, target: any) => {
      return am4core.color(props.data.colors[target.dataItem.index]);
    });

    chart.current = x;

    return () => {
      x.dispose();
    };
  }, [props.data.colors, props.data.data, props.id]);

  return <div id={props.id} className="h-100"></div>;
};

export default PieChart;
