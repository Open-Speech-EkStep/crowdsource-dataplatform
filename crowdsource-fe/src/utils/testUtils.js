import { render } from '@testing-library/react';

const Wrapper = ({ children }) => children;

const customRender = (ui, options) => render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };
