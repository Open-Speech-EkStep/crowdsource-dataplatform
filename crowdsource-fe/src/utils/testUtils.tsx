/* eslint-disable jest/no-export */
import type { ReactNode, ReactElement, ComponentType } from 'react';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import puppeteer from 'puppeteer';
import type { Browser } from 'puppeteer';

const Wrapper: ComponentType = ({ children }: { children?: ReactNode }) => children as ReactElement;

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };

export const verifyAxeTest = ({ container }: { container: RenderResult['container'] }) => {
  it('should not fail an axe audit', async () => {
    let axeResult;

    await act(async () => {
      axeResult = await axe(container);
    });

    expect(axeResult).toHaveNoViolations();
  });
};

export const verifyVRTest = (title: string, url: string) => {
  describe(`${title} VR test`, () => {
    let browser: Browser;

    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: true,
        devtools: false,
        ignoreHTTPSErrors: true,
        args: ['--start-fullscreen'],
      });
    });

    it('match snapshot', async () => {
      const page = await browser.newPage();
      await page.setViewport({ width: 1920, height: 1080 });

      await page.goto(url, {
        waitUntil: 'networkidle2',
      });

      await page.addStyleTag({
        content: `
      *,
      *::after,
      *::before {
          transition-delay: 0s !important;
          transition-duration: 0s !important;
          animation-delay: -0.0001s !important;
          animation-duration: 0s !important;
          animation-play-state: paused !important;
          caret-color: transparent !important;
      }
      `,
      });

      const image = await page.screenshot({
        fullPage: true,
      });

      expect(image).toMatchImageSnapshot();
    });

    afterAll(async () => {
      await browser.close();
    });
  });
};
