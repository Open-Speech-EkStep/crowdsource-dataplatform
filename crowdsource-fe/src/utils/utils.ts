import { jsPDF } from 'jspdf';
import { i18n } from 'next-i18next';

import { ErrorStatusCode } from 'constants/errorStatusCode';
import { INITIATIVES_MAPPING } from 'constants/initiativeConstants';
import { KEYBOARD_ERROR, LANGUAGE_UNICODE, OTHER_LANGUAGE_UNICODE } from 'constants/Keyboard';
import type { InitiativeType } from 'types/InitiativeType';

import apiPaths from '../constants/apiPaths';

var platform = require('platform');

export const convertIntoHrsFormat = (data: any, isSeconds = true) => {
  const hours = Math.floor(data / 3600);
  const remainingAfterHours = data % 3600;
  const minutes = Math.floor(remainingAfterHours / 60);
  const seconds = Math.round(remainingAfterHours % 60);
  if (isSeconds) {
    return { hours, minutes, seconds };
  } else {
    return { hours, minutes };
  }
};

export const roundOffValue = (average: any, precision: number) => {
  const formattedAverage = average.toFixed(precision);
  return formattedAverage;
};

export const capitalizeFirstLetter = (strValue: string) => {
  return strValue?.charAt(0).toUpperCase() + strValue?.slice(1);
};

const translateText = (text: string) => i18n?.t(text);

export const formatTime = (hours: number, minutes = 0, seconds = 0) => {
  const hrStr = translateText('hours1');
  const minStr = translateText('minutes1');
  const secStr = translateText('seconds1');
  let result = '';
  if (hours > 0) {
    result += `${hours} ${hrStr} `;
  }
  if (minutes > 0) {
    result += `${minutes} ${minStr} `;
  }
  if (hours === 0 && minutes === 0 && seconds > 0) {
    result += `${seconds} ${secStr} `;
  }

  if (hours === 0 && minutes === 0 && seconds === 0) {
    result += `0 ${secStr} `;
  }

  return result.substr(0, result.length - 1);
};

export const convertTimeFormat = (value: any) => {
  const { hours, minutes, seconds } = convertIntoHrsFormat(Number(value) * 60 * 60);
  const data = formatTime(hours, minutes, seconds);
  return data;
};

const roundDuration = (value: number) => {
  return Math.round(value * 100) / 100;
};

export const getHoursValue = (value: any) => {
  return roundDuration(value);
};

export const getMinutesValue = (value: any) => {
  return roundDuration(value * 60);
};

export const getHoursText = (value: any) => {
  return `${getHoursValue(value)} ${i18n?.t('hours2')}`;
};

export const getMinutesText = (value: any) => {
  return `${getMinutesValue(value)} ${i18n?.t('minutes2')}`;
};

export const isSunoOrBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['asr', 'text'];
  return boloOrSuno.includes(value);
};

export const isBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['text'];
  return boloOrSuno.includes(value);
};

export const findInputError = (text: string, currentModule: string, language: string) => {
  if (!text.length) return;
  const newText = text.replace(/\s/g, ''); //read input value, and remove "space" by replace \s
  let error: any = KEYBOARD_ERROR.language;
  const isSunoModule = currentModule === INITIATIVES_MAPPING.suno;
  const specialSymbols = /[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007E\u0964-\u0965]/;
  if (currentModule === INITIATIVES_MAPPING.suno && specialSymbols.test(newText) === true) {
    return KEYBOARD_ERROR.symbol;
  }
  Object.entries(isSunoModule ? LANGUAGE_UNICODE : OTHER_LANGUAGE_UNICODE).forEach(([key, value]) => {
    if (value.test(newText)) {
      //Check Unicode to see which one is true
      if (key.toLowerCase() === language.toLowerCase()) {
        error = KEYBOARD_ERROR.noError;
      }
    }
  });
  return error;
};

export const fetchLocationInfo = async () => {
  const result = await fetch('https://www.cloudflare.com/cdn-cgi/trace')
    .then(res => res.text())
    .then(async ipAddressText => {
      const dataArray = ipAddressText.split('\n');
      let ipAddress = '';
      for (let ind in dataArray) {
        if (dataArray[ind].startsWith('ip=')) {
          ipAddress = dataArray[ind].replace('ip=', '');
          break;
        }
      }
      if (ipAddress.length !== 0) {
        const data = await fetch(`${apiPaths.locationInfo}?ip=${ipAddress}`);
        return data && data.json();
      } else {
        return new Promise((resolve, reject) => {
          reject('Ip Address not available');
        });
      }
    });

  return result;
};

export const getDeviceInfo = () => {
  const os = platform.os;
  return os.family + ' ' + os.version;
};

export const getBrowserInfo = () => {
  return platform.name + ' ' + platform.version;
};

export const downloadBadge = (
  locale: string | undefined,
  initiative: string,
  source: string,
  badgeType: string,
  badgeId: string
) => {
  const pdf = new jsPDF();
  const img = new Image();
  img.src = `/images/${locale}/badges/${locale}_${initiative}_${badgeType}_${source}.png`;
  img.crossOrigin = 'Anonymous';
  img.onload = function () {
    pdf.addImage(img, 50, 10, 105, 130);
    pdf.save(`${badgeType}-badge.pdf`);
  };

  pdf.text(`Badge Id : ${badgeId}`, 36, 190);
};

export const groupBy = function (list: any, keyName: any) {
  if (!list || list.length === 0 || !list[0][keyName]) {
    return [];
  }
  return list.reduce(function (pair: any, key: any) {
    (pair[key[keyName]] = pair[key[keyName]] || []).push(key);
    return pair;
  }, {});
};

/* istanbul ignore next */
export const visualize = (visualizer: any, analyser: any) => {
  const canvasCtx = visualizer.getContext('2d');
  const bufferLength = analyser.frequencyBinCount;
  const dataArray = new Uint8Array(bufferLength);
  const WIDTH = visualizer.width;
  const HEIGHT = visualizer.height;
  // TODO do we need to limit the number of time visualize refreshes per second
  // so that it can run on Android processors without causing audio to drop?
  function draw() {
    // this is more efficient than calling with processor.onaudioprocess
    // and sending floatarray with each call...
    requestAnimationFrame(draw);
    analyser.getByteTimeDomainData(dataArray);
    // canvasCtx.fillStyle = 'rgb(255, 255, 255, 0)';
    // canvasCtx.fillRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.clearRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.lineWidth = 2;
    canvasCtx.strokeStyle = 'rgb(0,123,255)';
    canvasCtx.beginPath();
    const sliceWidth = (WIDTH * 1.0) / bufferLength;
    let x = 0;
    for (let i = 0; i < bufferLength; i++) {
      let v = dataArray[i] / 128.0; // uint8
      let y = (v * HEIGHT) / 2; // uint8
      if (i === 0) {
        canvasCtx.moveTo(x, y);
      } else {
        canvasCtx.lineTo(x, y);
      }
      x += sliceWidth;
    }
    canvasCtx.lineTo(visualizer.width, visualizer.height / 2);
    canvasCtx.stroke();
  }

  draw();
};

export const getErrorMsg = (error: any) => {
  if (error && error.status === ErrorStatusCode.TOO_MANY_REQUEST) {
    return 'multipleRequestApiError';
  } else {
    return 'apiFailureError';
  }
};
