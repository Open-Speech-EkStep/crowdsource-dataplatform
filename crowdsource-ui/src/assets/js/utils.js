const { HOUR_IN_SECONDS, SIXTY, ALL_LANGUAGES, CONTRIBUTION_LANGUAGE, DEFAULT_CON_LANGUAGE, ErrorStatusCode } = require("./constants");
const fetch = require('./fetch')
const platform = require('./platform');


function getDeviceInfo() {
  const os = platform.os;
  let info = "";
  if (os.family) {
    info = info + os.family;
  }
  if (os.version) {
    info = info + " " + os.version;
  }
  if (platform.product) {
    info = info + " " + platform.product;
  }
  return info.trim();
}

function getBrowserInfo() {
  let info = "";
  if (platform.name) {
    info = info + platform.name;
  }
  if (platform.version) {
    info = info + " " + platform.version;
  }
  return info.trim();
}

function setCookie(cname, cvalue, exdays) {
  var d = new Date();
  d.setTime(d.getTime() + (exdays * 24 * 60 * 60 * 1000));
  var expires = "expires=" + d.toGMTString();
  document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
}

function getCookie(cname) {
  var name = cname + "=";
  var decodedCookie = decodeURIComponent(document.cookie);
  var ca = decodedCookie.split(';');
  for (var i = 0; i < ca.length; i++) {
    var c = ca[i];
    while (c.charAt(0) == ' ') {
      c = c.substring(1);
    }
    if (c.indexOf(name) == 0) {
      return c.substring(name.length, c.length);
    }
  }
  return "";
}

function showElement(element) {
  element.removeClass('d-none');
}

function hideElement(element) {
  element.addClass('d-none');
}

function convertPXToVH(px) {
  return px * (100 / document.documentElement.clientHeight);
}

function setPageContentHeight() {
  const $footer = $('footer');
  const $nav = $('.navbar');
  const edgeHeightInPixel = $footer.outerHeight() + $nav.outerHeight()
  const contentHeightInVH = 100 - convertPXToVH(edgeHeightInPixel)
  $('#content-wrapper').css('min-height', contentHeightInVH + 'vh');
}

function toggleFooterPosition() {
  // const $footer = $('footer');
  // $footer.toggleClass('fixed-bottom')
  // $footer.toggleClass('bottom')
}

const safeJson = (res) => {
  if (!res) return res;
  if (typeof res.json === "function") return res.json();
  return res;
}

const safeJsonParse = (value, fallBack = {}) => {
  try {
    return JSON.parse(value);
  } catch (error) {
    return fallBack;
  }
}

function fetchLocationInfo() {
  //https://api.ipify.org/?format=json
  let regionName = localStorage.getItem("state_region") || "NOT_PRESENT";
  let countryName = localStorage.getItem("country") || "NOT_PRESENT";
  if (regionName !== "NOT_PRESENT" && countryName !== "NOT_PRESENT" && regionName.length > 0 && countryName.length > 0) {
    return new Promise((resolve) => {
      resolve({ "regionName": regionName, "country": countryName })
    })
  }
  return fetch('https://www.cloudflare.com/cdn-cgi/trace').then(res => res.text()).then(ipAddressText => {
    const dataArray = ipAddressText.split('\n');
    let ipAddress = "";
    for (let ind in dataArray) {
      if (dataArray[ind].startsWith("ip=")) {
        ipAddress = dataArray[ind].replace('ip=', '');
        break;
      }
    }
    if (ipAddress.length !== 0) {
      return fetch(`/location-info?ip=${ipAddress}`);
    } else {
      return new Promise((resolve, reject) => {
        reject("Ip Address not available")
      })
    }
  });
}

const showErrorPopup = (status = 500) => {
  const text = getErrorText(status);
  bindErrorText(text);
  const $errorDialog = $('#errorPopup');
  $errorDialog.modal('show');
}

const getErrorText = (status) => {
  const errorText = status === ErrorStatusCode.SERVICE_UNAVAILABLE ?
    translate("We are processing multiple requests at the moment. Please try again after sometime.")
    : translate("An unexpected error has occurred.");
  return errorText;
}

const bindErrorText = (text) => {
  const $errorText = $("#error-text");
  $errorText.text(text);
}

const performAPIRequest = (url) => {
  return fetch(url, {
    credentials: 'include',
    mode: 'cors'
  }).then((data) => {
    if (!data.ok) {
      throw (data.status || 500);
    } else {
      return Promise.resolve(data.json());
    }
  })
  .catch(errStatus => {
    showErrorPopup(errStatus);
    throw errStatus
  });
}

const getLocaleString = function () {
  return new Promise(function (resolve, reject) {
    const locale = sessionStorage.getItem("i18n") ?? "en";
    performAPIRequest(`/get-locale-strings/${locale}`)
      .then((response) => {
        if (response) {
          localStorage.setItem('localeString', JSON.stringify(response));
          resolve(response);
        }
      }).catch((err) => { reject(err) });
  });
}

const updateLocaleLanguagesDropdown = (language) => {
  const dropDown = $('#localisation_dropdown');
  language = localStorage.getItem(CONTRIBUTION_LANGUAGE) || DEFAULT_CON_LANGUAGE;
  const localeLang = ALL_LANGUAGES.find(ele => ele.value.toLowerCase() === language.toLowerCase());
  if (language.toLowerCase() === "english" || localeLang.hasLocaleText === false) {
    dropDown.html('<a id="english" class="dropdown-item d-flex align-items-center py-3 py-md-2 py-lg-2" href="#" locale="en">English</a>');
  } else {
    dropDown.html(`<a id="english" class="dropdown-item py-3 py-md-2 py-lg-2" href="#" locale="en">English</a>
        <a id=${localeLang.value} class="dropdown-item d-flex align-items-center py-3 py-md-2 py-lg-2" href="#" locale="${localeLang.id}">${localeLang.text}</a>`);
  }
}

const calculateTime = function (totalSeconds, isSeconds = true) {
  const hours = Math.floor(totalSeconds / HOUR_IN_SECONDS);
  const remainingAfterHours = totalSeconds % HOUR_IN_SECONDS;
  const minutes = Math.floor(remainingAfterHours / SIXTY);
  const seconds = Math.round(remainingAfterHours % SIXTY);
  if (isSeconds) {
    return { hours, minutes, seconds };
  } else {
    return { hours, minutes };
  }
};

const translate = function (string) {
  const localeStrings = JSON.parse(localStorage.getItem('localeString'));
  return localeStrings[string] || string;
}

const formatTime = function (hours, minutes = 0, seconds = 0, translate = true) {
  const localeStrings = JSON.parse(localStorage.getItem('localeString') || "{}");
  const hrStr = translate ? localeStrings['hour(s)'] || 'hour(s)' : 'hour(s)';
  const minStr = translate ? localeStrings['minute(s)'] || 'minute(s)' : 'minute(s)';
  const secStr = translate ? localeStrings['second(s)'] || 'second(s)' : 'second(s)';
  let result = "";
  if (hours > 0) {
    result += `${hours} ${hrStr} `;
  }
  if (minutes > 0) {
    result += `${minutes} ${minStr} `;
  }
  if (hours === 0 && minutes === 0 && seconds > 0) {
    result += `${seconds} ${secStr} `;
  }

  if (hours === 0 && minutes === 0 && seconds === 0) {
    result += `0 ${secStr} `;
  }

  if (result.charAt(result.length - 1) !== ' ')
    return result.substr(0, result.length);
  else
    return result.substr(0, result.length - 1);
};

const formatTimeForLegends = function (hours, minutes = 0, seconds = 0, isLabelRequired = true) {
  const localsStrings = JSON.parse(localStorage.getItem('localeString'));
  const hrStr = localsStrings['hours'];
  const minStr = localsStrings['minutes'];
  const secStr = localsStrings['seconds'];
  if (hours && minutes) {
    return isLabelRequired ? `${hours}.${minutes} ${hrStr}` : `${hours}.${minutes}`;
  }
  if (hours == 0 && minutes == 0) {
    return isLabelRequired ? `${seconds} ${secStr}` : `${seconds}`;
  }
  const hoursStr = hours ? (isLabelRequired ? `${hours} ${hrStr}` : `${hours}`) : '';
  const minutesStr = minutes ? (isLabelRequired ? `${minutes} ${minStr}` : `${minutes}`) : '';
  return `${hoursStr} ${minutesStr}`.trim();
}

const setFooterPosition = () => {
  const contentHeight = $('#page-content').outerHeight();
  const bodyHeight = $('body').outerHeight();
  const navHeight = $('nav').outerHeight();
  const footerHeight = $('footer').outerHeight();
  const totalHeight = contentHeight + navHeight + footerHeight;
  if (bodyHeight <= totalHeight) {
    $('footer').removeClass('fixed-bottom').addClass('bottom');
  }
}

const reportSentenceOrRecording = (reqObj) => {
  return new Promise(function (resolve, reject) {
    try {
      fetch('/report', {
        method: "POST",
        credentials: 'include',
        mode: 'cors',
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(reqObj),
      })
        .then((res) => res.json())
        .then((resp) => {
          resolve(resp);
        })
    } catch (err) {
      reject(err);
    }
  });
}

const getJson = (path) => {
    return new Promise((resolve) => {
      $.getJSON(path, (data) => {
        resolve(data);
      }).fail((e) => {
        if(e && e.statusText !== "error") {
          showErrorModal();
        }
      });
    })
}

const covertStringToCapitalised = (tempStr) => {
  if (tempStr) {
    return tempStr.substr(0, 1).toUpperCase() + tempStr.substr(1);
  } else {
    return tempStr;
  }
}

const showErrorModal = () => {
  const $errorText = $("#error-text");
  $errorText.text("");
  $errorText.text(translate("An unexpected error has occurred."));
  const $errorDialog = $('#errorPopup');
  $errorDialog.modal('show');
}

const getLanguageBadge = (contibutedLanguage, badgeType, source, initiativeType) => {
  const language = ALL_LANGUAGES.find(language => language.value.toLowerCase() === contibutedLanguage.toLowerCase());
  const langaugePrefix = language ? language.id : 'en';
  return `/img/${langaugePrefix}_${initiativeType}_${badgeType}_${source}.svg`;
}


module.exports = {
  safeJsonParse,
  safeJson,
  toggleFooterPosition,
  fetchLocationInfo,
  updateLocaleLanguagesDropdown,
  calculateTime,
  covertStringToCapitalised,
  formatTime,
  getLocaleString,
  performAPIRequest,
  showElement,
  hideElement,
  setFooterPosition,
  reportSentenceOrRecording,
  getCookie,
  setCookie,
  getJson,
  setPageContentHeight,
  getDeviceInfo,
  getBrowserInfo,
  formatTimeForLegends,
  getLanguageBadge,
  translate
}
