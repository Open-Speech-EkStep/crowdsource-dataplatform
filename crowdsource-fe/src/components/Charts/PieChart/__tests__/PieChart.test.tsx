import { render } from 'utils/testUtils';

import PieChart from '../PieChart';

describe('PieChart', () => {
  const setup = () =>
    render(
      <PieChart
        id="chart_id"
        data={{
          data: [
            { category: 'a', value: 10 },
            { category: 'b', value: 7 },
          ],
        }}
      />
    );

  it('should render the component and matches it against stored snapshot', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});
