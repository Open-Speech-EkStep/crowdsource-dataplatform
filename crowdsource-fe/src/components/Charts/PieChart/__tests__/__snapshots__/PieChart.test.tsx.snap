// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    id="chart_id"
    style="width: 100%; height: 500px;"
  >
    <div
      style="width: 100%; height: 100%; position: relative;"
    >
      <svg
        role="group"
        style="width: 100%; height: 100%; overflow: visible;"
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs>
          <clippath
            id="id-45"
          >
            <rect
              height="0"
              width="0"
            />
          </clippath>
          <filter
            filterUnits="objectBoundingBox"
            height="200%"
            id="filter-id-50"
            width="200%"
            x="-50%"
            y="-50%"
          >
            <fegaussianblur
              in="SourceGraphic"
              result="blurOut"
              stdDeviation="1.5"
            />
            <feoffset
              dx="1"
              dy="1"
              result="offsetBlur"
            />
            <feflood
              flood-color="#000000"
              flood-opacity="0.5"
            />
            <fecomposite
              in2="offsetBlur"
              operator="in"
            />
            <femerge>
              <femergenode />
              <femergenode
                in="SourceGraphic"
              />
            </femerge>
          </filter>
          <filter
            filterUnits="objectBoundingBox"
            height="120%"
            id="filter-id-61"
            width="120%"
            x="-10%"
            y="-10%"
          >
            <fecolormatrix
              type="saturate"
              values="0"
            />
          </filter>
          <lineargradient
            id="gradient-id-64"
            x1="1%"
            x2="99%"
            y1="59%"
            y2="41%"
          >
            <stop
              offset="0"
              stop-color="#474758"
            />
            <stop
              offset="0.75"
              stop-color="#474758"
              stop-opacity="1"
            />
            <stop
              offset="0.755"
              stop-color="#3cabff"
              stop-opacity="1"
            />
          </lineargradient>
          <clippath
            id="id-84"
          />
          <filter
            filterUnits="objectBoundingBox"
            height="200%"
            id="filter-id-89"
            width="200%"
            x="-50%"
            y="-50%"
          >
            <fegaussianblur
              in="SourceGraphic"
              result="blurOut"
              stdDeviation="1.5"
            />
            <feoffset
              dx="1"
              dy="1"
              result="offsetBlur"
            />
            <feflood
              flood-color="#000000"
              flood-opacity="0.5"
            />
            <fecomposite
              in2="offsetBlur"
              operator="in"
            />
            <femerge>
              <femergenode />
              <femergenode
                in="SourceGraphic"
              />
            </femerge>
          </filter>
        </defs>
        <g>
          <g
            fill="#ffffff"
            fill-opacity="0"
          >
            <rect />
          </g>
          <g>
            <g
              clip-path="url(\\"#id-45\\")"
              role="region"
            >
              <g>
                <g>
                  <g>
                    <g>
                      <g>
                        <g>
                          <g>
                            <g
                              role="group"
                            >
                              <g>
                                <g
                                  clip-path="url(\\"#id-84\\")"
                                >
                                  <g />
                                </g>
                                <g />
                                <g>
                                  <g />
                                </g>
                                <g>
                                  <g />
                                </g>
                                <g>
                                  <g />
                                </g>
                              </g>
                            </g>
                          </g>
                        </g>
                        <g>
                          <g>
                            <g>
                              <g />
                            </g>
                          </g>
                        </g>
                      </g>
                    </g>
                    <g
                      role="group"
                    >
                      <g />
                    </g>
                  </g>
                </g>
              </g>
            </g>
            <g>
              <g>
                <g
                  opacity="0"
                  role="tooltip"
                  visibility="hidden"
                >
                  <g
                    fill="#ffffff"
                    fill-opacity="0.9"
                    filter="url(\\"#filter-id-50\\")"
                    stroke="#ffffff"
                    stroke-opacity="1"
                    stroke-width="1"
                    style="pointer-events: none;"
                  >
                    <path />
                  </g>
                  <g>
                    <g
                      fill="#ffffff"
                      style="pointer-events: none;"
                    >
                      <g
                        transform="translate(0,7)"
                      />
                    </g>
                  </g>
                </g>
                <g
                  visibility="hidden"
                >
                  <g
                    fill="#ffffff"
                    opacity="1"
                  >
                    <rect />
                  </g>
                  <g>
                    <g>
                      <g>
                        <g
                          fill="#f3f3f3"
                          fill-opacity="0.8"
                          stroke-opacity="1"
                        >
                          <g>
                            <g />
                          </g>
                        </g>
                        <g
                          fill="#000000"
                          fill-opacity="0.2"
                          stroke-opacity="1"
                        >
                          <g>
                            <g />
                          </g>
                        </g>
                        <g
                          fill="#000000"
                          fill-opacity="0.4"
                        >
                          <g />
                        </g>
                      </g>
                    </g>
                  </g>
                </g>
                <g
                  aria-labelledby="id-61-title"
                  filter="url(\\"#filter-id-61\\")"
                  opacity="0.3"
                  style="cursor: pointer;"
                >
                  <g
                    fill="#ffffff"
                    opacity="0"
                  >
                    <rect />
                  </g>
                  <g>
                    <g
                      fill="none"
                      shape-rendering="auto"
                      stroke="#3cabff"
                      stroke-opacity="1"
                      stroke-width="1.7999999999999998"
                    >
                      <path
                        d=" M15,15  C17.4001,15 22.7998,15.0001 27,15 C31.2002,14.9999 33.2999,6 36,6 C38.7001,6 38.6999,10.5 40.5,10.5 C42.3001,10.5 42.2999,6 45,6 C47.7001,6 50.9999,14.9999 54,15 C57.0002,15.0001 58.7999,15 60,15"
                      />
                    </g>
                    <g
                      fill="none"
                      shape-rendering="auto"
                      stroke="url(\\"#gradient-id-64\\")"
                      stroke-opacity="1"
                      stroke-width="1.7999999999999998"
                    >
                      <path
                        d=" M6,15  C8.2501,15 9.7498,15.0001 15,15 C20.2502,14.9999 20.7748,3.6 27,3.6 C33.2252,3.6 33.8998,14.9999 39.9,15 C45.9002,15.0001 45.9748,15 51,15 C56.0252,15 57.7499,15 60,15"
                      />
                    </g>
                  </g>
                  <title
                    id="id-61-title"
                  >
                    Chart created using amCharts library
                  </title>
                </g>
                <g
                  opacity="0"
                  role="tooltip"
                  visibility="hidden"
                >
                  <g
                    fill="#ffffff"
                    fill-opacity="0.9"
                    filter="url(\\"#filter-id-89\\")"
                    stroke="#ffffff"
                    stroke-opacity="1"
                    stroke-width="1"
                    style="pointer-events: none;"
                  >
                    <path />
                  </g>
                  <g>
                    <g
                      fill="#ffffff"
                      style="pointer-events: none;"
                    >
                      <g
                        transform="translate(0,7)"
                      />
                    </g>
                  </g>
                </g>
              </g>
            </g>
          </g>
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
